sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/XmlExternalEntityShouldNotBeParsed.cs

```diff
@@ -0,0 +1,187 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics.CodeAnalysis;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class XmlExternalEntityShouldNotBeParsed : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S2755";
+        private const string MessageFormat = "Disable access to external entities in XML parsing.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static ImmutableArray<KnownType> UnsafeXmlResolvers { get; } = ImmutableArray.Create(
+                KnownType.System_Xml_XmlUrlResolver,
+                KnownType.System_Xml_Resolvers_XmlPreloadedResolver
+        );
+
+        // The value of System.Xml.DtdProcessing.Parse
+        private const int DtdProcessingParse = 2;
+
+        // For the XXE rule we actually need to know about .NET 4.5.2,
+        // but it is good enough given the other .NET 4.x do not have support anymore
+        private INetFrameworkVersionProvider VersionProvider;
+
+        public XmlExternalEntityShouldNotBeParsed()
+            : this(NetFrameworkVersionProvider.Instance)
+        {
+        }
+
+        internal /*for testing*/ XmlExternalEntityShouldNotBeParsed(INetFrameworkVersionProvider netFrameworkVersionProvider)
+        {
+            VersionProvider = netFrameworkVersionProvider;
+        }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            (new XmlDocumentShouldBeSafe(VersionProvider)).InternalInitialize(context);
+            (new XmlTextReaderShouldBeSafe(VersionProvider)).InternalInitialize(context);
+
+            // FIXME Implement for XmlReader and XPathNavigator
+         }
+
+        [SuppressMessage("AnalyzerCorrectness", "RS1001:Missing diagnostic analyzer attribute.", Justification = "This is not a different diagnostic.")]
+        private class XmlDocumentShouldBeSafe : ObjectShouldBeInitializedCorrectlyBase
```

Comment:
Currently `XmlDocumentShouldBeSafe` shares common behavior with `XmlTextReaderShouldBeSafe`. Would make sense to have a common base class in order to reduce the duplication? 
Something similar with:
```
private abstract class XmlExternalEntityBase : ObjectShouldBeInitializedCorrectlyBase
{
    protected NetFrameworkVersion NetFrameworkVersion;

    protected readonly INetFrameworkVersionProvider VersionProvider;

    internal XmlExternalEntityBase(INetFrameworkVersionProvider netFrameworkVersionProvider)
    {
        VersionProvider = netFrameworkVersionProvider;
    }

    protected override void CompilationAction(Compilation compilation)
    {
        NetFrameworkVersion = VersionProvider.GetDotNetFrameworkVersion(compilation);
    }

    // FIXME create constructor that gets the target framework and IsAllowedValue should consider that
    protected override bool IsAllowedValue(ISymbol symbol) =>
        !IsUnsafeXmlResolverConstructor(symbol) &&
        !symbol.GetSymbolType().IsAny(UnsafeXmlResolvers);
}
```